File: elixir/find-champion-ii.ex
defmodule Solution do
  @spec find_champion(n :: integer, edges :: [[integer]]) :: integer
  def find_champion(1, _), do: 0
  def find_champion(_, []), do: -1
  def find_champion(n, edges) do
    stronger_than = Map.new(Enum.map(edges, fn [stronger, weaker] -> {weaker, stronger} end))
    if map_size(stronger_than) == n - 1 do
      find_strongest(stronger_than, edges |> hd |> hd)
    else
      -1
    end
  end

  defp find_strongest(map, start) do
    if Map.has_key?(map, start) do
      find_strongest(map, Map.get(map, start))
    else
      start
    end
  end
end

File: javascript/find-champion-ii.js
/**
 * @param {number} n
 * @param {number[][]} edges
 * @return {number}
 */
function findChampion(n, edges) {
    const weak = new Set();
    for (const [_stronger, weaker] of edges) {
        weak.add(weaker);
    }
    if (weak.size == n - 1) {
        return -1;
    }
    for (let i = 0; i < n; i++) {
        if (!weak.has(i)) {
            return i;
        }
    }
}
